networkx/algorithms/dag.py

```diff
@@ -79,172 +82,184 @@ def is_directed_acyclic_graph(G):
+    topologically_sorted_nodes : iterable
+        An iterable of node names in topological sorted order.
 
     Raises
     ------
     NetworkXError
-        Topological sort is defined for directed graphs only. If the
-        graph G is undirected, a NetworkXError is raised.
+        Topological sort is defined for directed graphs only. If the graph G
+        is undirected, a NetworkXError is raised.
 
     NetworkXUnfeasible
-        If G is not a directed acyclic graph (DAG) no topological sort
-        exists and a NetworkXUnfeasible exception is raised.
+        If G is not a directed acyclic graph (DAG) no topological sort exists
+        and a NetworkXUnfeasible exception is raised.  This can also be
+        raised if G is changed while the returned iterator is being processed.
+
+    RuntimeError
+        If G is changed while the returned iterator is being processed.
 
     Notes
     -----
     This algorithm is based on a description and proof in
-    The Algorithm Design Manual [1]_ .
+    Introduction to algorithms - a creative approach [1]_ .
 
     See also
     --------
-    is_directed_acyclic_graph
+    is_directed_acyclic_graph, lexicographical_topological_sort
 
     References
     ----------
-    .. [1] Skiena, S. S. The Algorithm Design Manual  (Springer-Verlag, 1998). 
-        http://www.amazon.com/exec/obidos/ASIN/0387948600/ref=ase_thealgorithmrepo/
+    .. [1] Manber, U. (1989). Introduction to algorithms - a creative approach. Addison-Wesley.
+        http://www.amazon.com/Introduction-Algorithms-A-Creative-Approach/dp/0201120372
     """
     if not G.is_directed():
         raise nx.NetworkXError(
             "Topological sort not defined on undirected graphs.")
 
-    # nonrecursive version
-    seen = set()
-    order = []
-    explored = set()
-
-    if nbunch is None:
-        nbunch = G.nodes_iter()
-    for v in nbunch:     # process all vertices in G
-        if v in explored:
-            continue
-        fringe = [v]   # nodes yet to look at
-        while fringe:
-            w = fringe[-1]  # depth first search
-            if w in explored:  # already looked down this branch
-                fringe.pop()
-                continue
-            seen.add(w)     # mark as seen
-            # Check successors for cycles and for new nodes
-            new_nodes = []
-            for n in G[w]:
-                if n not in explored:
-                    if n in seen:  # CYCLE !!
-                        raise nx.NetworkXUnfeasible("Graph contains a cycle.")
-                    new_nodes.append(n)
-            if new_nodes:   # Add new_nodes to fringe
-                fringe.extend(new_nodes)
-            else:           # No new nodes so w is fully explored
-                explored.add(w)
-                order.append(w)
-                fringe.pop()    # done considering this node
-    if reverse:
-        return order
-    else:
-        return list(reversed(order))
+    indegree_map = G.in_degree()
```

Comment:
Maybe simpler, to remove key deletion and collapse four lines of code into two:

    indegree = {v: d for v, d in G.in_degree() if d > 0}
    indegree_zero = [v for v, d in G.in_degree() if d == 0]

(Assuming that `in_degree` returns an iterator over pairs, as it will once some current pull requests get merged.)